trunk/src/meshmod/close_mesh.F90

```diff
@@ -203,3 +215,213 @@ subroutine close_mesh()
+#if DEBUG_MODE
+   if ((RANK.eq.ROOT) .and. (iprint.eq.2)) then
+      write(*,*) 'close_mesh: RESOLVE THE NODES MORE THAN ONE IRREGULARITY'
+   endif
+#endif
+!
+!---------------------------------------------------------
+!     Step 2 : pick the middle nodes to be refined
+!---------------------------------------------------------
+!
+!$OMP PARALLEL DO                                     &
+!$OMP PRIVATE(mdle,nodesl,norientl,nod,ntype,nflag,   &
+!$OMP         nflag,krefe,kreff,kref,j,nv,ne,nf,nve)
+      do i=1,NRELES_SUBD
+         mdle=ELEM_SUBD(i)
+         call elem_nodes(mdle, nodesl,norientl)
+!
+         ntype = NODES(mdle)%ntype
+         nflag = .FALSE.
+!
+!        check edges
+!        ~~~~~~~~~~~~
+         krefe = 0
+!
+         nv = nvert(ntype)
+         ne = nedge(ntype)
+         nf = nface(ntype)
+         nve = nv + ne
+!
+         do j = 1,ne
+            nod = nodesl(nv+j)
+            if (visited(nod)) then
+               krefe(j)=1
+               nflag=.TRUE.
+            endif
+         enddo
+!
+!        check faces
+!        ~~~~~~~~~~~~
+         kreff = 0
+         do j = 1,nf
+            nod = nodesl(nve+j)
+            if (visited(nod)) then
+               call get_isoref(nod, kreff(j))
+               nflag=.TRUE.
+            endif
+         enddo
+!
+!        find out more than one-irregular
+!        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+         if (nflag) then
+!
+!$OMP CRITICAL
+#if DEBUG_MODE
+            if ((RANK.eq.ROOT) .and. (iprint.eq.1)) then
+               ne = nedge(NODES(mdle)%ntype)
+               nf = nface(NODES(mdle)%ntype)
+               write(*,*) 'close_mesh: mdle = ', mdle
+               write(*,7003) krefe(1:12)
+   7003        format('krefe = ',12i2)
+               write(*,7004) kreff(1:6)
+   7004        format('kreff = ',6i3)
+               call pause
+            endif
+#endif
+!           increment counter
+            ic=ic+1
+!
+!           find out proper refinement flag
+            if (is_iso_only()) then
+               call get_isoref(mdle, kref)
+            else
+!           ...-------------------------------------------------------------------
+!           ...Option 1: do minimum refinement that is necessary
+!               call find_element_closing_ref(ntype,kreff,krefe, kref)
+!           ...-------------------------------------------------------------------
+!           ...Option 2: always ask for isotropic refinement
+               call get_isoref(mdle, kref)
+!           ...-------------------------------------------------------------------
+!           ...Option 3: always ask for radial (xy) refinement (FIBER LASER)
+!               select case (NODES(mdle)%ntype)
+!                  case(MDLB); kref = 110
+!                  case(MDLP); kref = 10
+!               end select
+!           ...-------------------------------------------------------------------
+            endif
+            list(1,ic) = mdle
+            list(2,ic) = kref
+!$OMP END CRITICAL
```

Comment:
Reduction on `ic` wouldn't work; that makes a local copy of `ic`, sums it up locally, then combines the results at the end.

Reply:
So threads would try to assign in the same spot during the loop

Reply:
Oh, I see, you are writing the list here using the `ic` increment unlike in the sequential `close_mesh`. Since the list is being allocated on size `NR_ELES_SUBD` anyway, I think the better option is to write the list using `i` w/o any critical section and then after the loop rewrite the list into consecutive numbers. Something like
```
j = 0
do i=1,NRELES_SUBD
   if (list(1,i) .eq. 0) cycle
   j = j+1
   list(1:2,j) = list(1:2,i)
enddo
#if DEBUG_MODE
if (j .ne. ic) then
   ! .. error
endif
#endif
```

Reply:
Not sure about performance difference, it'll probably depend on the number of marked elements, number of threads, etc. I can leave the `critical` but I'll make the section smaller, it doesn't need to include computing the closing refinement flag.

Reply:
Sure, that's fine. `close_mesh` isn't a bottleneck anymore, it's pretty quick now that it (and refresh) scale.

Reply:
That's great to hear, it was significant bottleneck before parallelization for sure.